src/Update.cpp

```diff
@@ -355,48 +326,26 @@ void DoDeath_FromCriticalorSARIorILI(int ai, int tn)
 	Person* a = Hosts + ai;
 	if (P.DoSeverity)
 	{
-		if (a->Severity_Current == Severity_Critical)
+		switch(a->Severity_Current)
 		{
-			StateT[tn].Critical--;
-			StateT[tn].Critical_age[HOST_AGE_GROUP(ai)]--;
-			StateT[tn].cumDeath_Critical++;
-			StateT[tn].cumDeath_Critical_age[HOST_AGE_GROUP(ai)]++;
-			if (P.DoAdUnits)
-			{
-				StateT[tn].Critical_adunit			[Mcells[a->mcell].adunit]--;
-				StateT[tn].cumDeath_Critical_adunit	[Mcells[a->mcell].adunit]++;
-			}
-			//// change current status (so that flags work if function called again for same person). Don't move this outside of this if statement, even though it looks like it can be moved safely. It can't.
-			a->Severity_Current = Severity_Dead;
-		}
-		else if (a->Severity_Current == Severity_SARI)
-		{
-			StateT[tn].SARI--;
-			StateT[tn].SARI_age[HOST_AGE_GROUP(ai)]--;
-			StateT[tn].cumDeath_SARI++;
-			StateT[tn].cumDeath_SARI_age[HOST_AGE_GROUP(ai)]++;
-			if (P.DoAdUnits)
-			{
-				StateT[tn].SARI_adunit			[Mcells[a->mcell].adunit]--;
-				StateT[tn].cumDeath_SARI_adunit	[Mcells[a->mcell].adunit]++;
-			}
-			//// change current status (so that flags work if function called again for same person). Don't move this outside of this if statement, even though it looks like it can be moved safely. It can't.
-			a->Severity_Current = Severity_Dead;
-		}
-		else if (a->Severity_Current == Severity_ILI)
-		{
-			StateT[tn].ILI--;
-			StateT[tn].ILI_age[HOST_AGE_GROUP(ai)]--;
-			StateT[tn].cumDeath_ILI++;
-			StateT[tn].cumDeath_ILI_age[HOST_AGE_GROUP(ai)]++;
-			if (P.DoAdUnits)
-			{
-				StateT[tn].ILI_adunit			[Mcells[a->mcell].adunit]--;
-				StateT[tn].cumDeath_ILI_adunit	[Mcells[a->mcell].adunit]++;
-			}
-			//// change current status (so that flags work if function called again for same person). Don't move this outside of this if statement, even though it looks like it can be moved safely. It can't.
-			a->Severity_Current = Severity_Dead;
+			case Severity_Critical:
+				FromCritical(tn, a->mcell, ai);
+				ToDeathCritical(tn, a->mcell, ai);
+				break;
+
+			case Severity_SARI:
+				FromSARI(tn, a->mcell, ai);
+				ToDeathSARI(tn, a->mcell, ai);
+				break;
+
+			case Severity_ILI:
+				FromILI(tn, a->mcell, ai);
+				ToDeathILI(tn, a->mcell, ai);
+				break;
 		}
+
+		//// change current status (so that flags work if function called again for same person). Don't move this outside of this if statement, even though it looks like it can be moved safely. It can't.
```

Comment:
Thank you @mstevens-uk for this. Overall, it tidies up the movement between the severity states well, and is appreciated greatly. However, there is a warning explicitly not to do this: please put the assignment `a->Severity_Current = Severity_Dead;` back into the individual cases of this switch. The switch does not consider all severity cases (i.e. is not exhaustive), so in rare cases `DoDeath_FromCriticalorSARIorILI` can be called before a person has had their severity assigned. Therefore they will die before they onset with symptoms - this causes all kinds of problems with bookkeeping (total deaths don't add up). It will pass regression tests because the ILI CFR is set to impossible in the example pre-parameters, but this won't always be so, and because ILI cases are so common this can matter a great deal. An identical problem can happen with `DoRecover_FromSeverity`  but I can see that this part of your code changes is fine.

Reply:
@dlaydon Thanks for your review. Yes, good spot. I've put the assignment back into the switch cases, and added the extra detail from your review into the comments.